src/tracing/metric_macros.h

```diff
@@ -27,13 +27,258 @@
+        if (!::chip::ChipError::IsSuccess(__err))                                                                                  \
+        {                                                                                                                          \
+            MATTER_LOG_METRIC(kMetricKey, __err);                                                                                  \
+            return;                                                                                                                \
+        }                                                                                                                          \
+    } while (false)
+
+/**
+ *  @def VerifyOrReturnWithMetric(kMetricKey, expr, ...)
+ *
+ *  @brief
+ *    Returns from the void function if expression evaluates to false. If the expression evaluates
+ *    to false, a metric with the specified key is emitted.
+ *
+ *  Example usage:
+ *
+ *  @code
+ *    VerifyOrReturnWithMetric(kMetricKey, param != nullptr, LogError("param is nullptr"));
+ *  @endcode
+ *
+ *  @param[in]  kMetricKey  Metric key for the metric event to be emitted if the expr evaluates
+ *                          to false. Value of the metric is set to false.
+ *  @param[in]  expr        A Boolean expression to be evaluated.
+ *  @param[in]  ...         Statements to execute before returning. Optional.
+ */
+#define VerifyOrReturnWithMetric(kMetricKey, expr, ...)                                                                            \
+    do                                                                                                                             \
+    {                                                                                                                              \
+        if (!(expr))                                                                                                               \
+        {                                                                                                                          \
+            MATTER_LOG_METRIC(kMetricKey, false);                                                                                  \
+            __VA_ARGS__;                                                                                                           \
+            return;                                                                                                                \
+        }                                                                                                                          \
+    } while (false)
+
+/**
+ *  @def VerifyOrReturnErrorWithMetric(kMetricKey, expr, code, ...)
+ *
+ *  @brief
+ *    Returns a specified error code if expression evaluates to false. If the expression evaluates
+ *    to false, a metric with the specified key is emitted with the value set to the code.
```

Comment:
"to the error code"?  Or "to code"?  Either would be better than what we have here.